img2dataset/downloader.py

```diff
@@ -34,6 +36,69 @@ def is_disallowed(headers, user_agent_token, disallowed_header_directives):
     return False
 
 
+class RobotFileParser(urllib.robotparser.RobotFileParser):
+    """Extends RobotFileParser to handle Request urls, and to parse responses more in line with RFC9309"""
+
+    def __init__(self, url=""):
+        super().__init__()
+        self.set_url(url)
+
+    def set_url(self, url):
+        """Sets the URL referring to a robots.txt file, from either a url string or a Request object"""
+        self.url = url
+        if isinstance(url, urllib.request.Request):
+            self.host = url.host
+            self.path = url.selector
+        else:
+            self.host, self.path = urllib.parse.urlparse(url)[1:3]
+
+    def read(self):
+        """Reads the robots.txt URL and feeds it to the parser."""
+        try:
+            with urllib.request.urlopen(self.url, timeout=10) as f:
+                if f.info().get_content_type() != "text/plain":
+                    # server did not respond with a parsable robots.txt file, handle the same as unavailable status
+                    self.allow_all = True
+                else:
+                    raw = f.read()
+                    self.parse(raw.decode("utf-8").splitlines())
+        except urllib.error.HTTPError as err:
+            if err.code in (401, 403) or 500 <= err.code <= 599:
```

Comment:
according to 2.3.1.3 of https://datatracker.ietf.org/doc/rfc9309/  401 and 403 are considered Unavailable hence should be allow all

Reply:
This was based on the cpython implementation of read, which treats authentication errors as unreachable statuses. I forgot to comment that behaviour so have added it now.